File: hash/swift/RosettaSquintHash/Sources/RosettaSquintHash/Hash.swift
import Foundation

/// Input image type for all hash algorithms.
public struct RGBImage: Equatable {
	public let width: Int
	public let height: Int
	public let data: [UInt8]
	public let channels: Channels

	public enum Channels: Equatable {
		case rgb   // 3 bytes per pixel
		case rgba  // 4 bytes per pixel; alpha composited against opaque black internally

		public var bytesPerPixel: Int {
			switch self {
			case .rgb:  return 3
			case .rgba: return 4
			}
		}
	}

	public init(width: Int, height: Int, data: [UInt8], channels: Channels) {
		self.width = width
		self.height = height
		self.data = data
		self.channels = channels
	}

	/// Validates that the data buffer length matches width * height * bytesPerPixel
	/// and that dimensions are positive. Throws `ImageHashError.shapeMismatch` on
	/// mismatch. Every public hash function calls this at its entry point.
	public func validate() throws {
		guard width > 0, height > 0 else {
			throw ImageHashError.shapeMismatch(
				lhs: ImageHashError.ShapeKey(height, width),
				rhs: ImageHashError.ShapeKey(0, 0)
			)
		}
		// Compute width * height * bytesPerPixel with explicit overflow checks
		// instead of `*`, which would trap on Int overflow (Swift Int is signed
		// and arithmetic traps in debug AND release builds via overflow operators).
		// A pathological caller can fabricate (width, height) outside the decode
		// pipeline; surface a typed shapeMismatch rather than SIGABRT.
		let (wh, owh) = width.multipliedReportingOverflow(by: height)
		guard !owh else {
			throw ImageHashError.shapeMismatch(
				lhs: ImageHashError.ShapeKey(data.count, 1),
				rhs: ImageHashError.ShapeKey(Int.max, 1)
			)
		}
		let (whc, owhc) = wh.multipliedReportingOverflow(by: channels.bytesPerPixel)
		guard !owhc else {
			throw ImageHashError.shapeMismatch(
				lhs: ImageHashError.ShapeKey(data.count, 1),
				rhs: ImageHashError.ShapeKey(Int.max, 1)
			)
		}
		guard data.count == whc else {
			// Encode expected length in the rhs ShapeKey (height carries expected total bytes)
			throw ImageHashError.shapeMismatch(
				lhs: ImageHashError.ShapeKey(data.count, 1),
				rhs: ImageHashError.ShapeKey(whc, 1)
			)
		}
	}
}

/// All errors thrown by the public API.
public enum ImageHashError: Error, Equatable {
	case invalidHashSize(Int)
	case notPowerOfTwo(Int)
	case hashSizeTooLarge(level: Int, maxLevel: Int)
	case invalidBinbits(Int)
	case invalidHex(String)
	case shapeMismatch(lhs: ShapeKey, rhs: ShapeKey)
	case emptyBits
	case nonRectangular

	/// Helper struct because Swift doesn't auto-synthesize Equatable for enum
	/// cases with tuple associated values.
	public struct ShapeKey: Equatable {
		public let height: Int
		public let width: Int
		public init(_ height: Int, _ width: Int) {
			self.height = height
			self.width = width
		}
	}
}

/// Hash value: a 2-D boolean grid (typically NxN, but 14xB for colorhash).
public struct Hash: Equatable, Hashable, CustomStringConvertible {
	internal let bits: [[Bool]]

	public init(bits: [[Bool]]) throws {
		guard !bits.isEmpty, !bits[0].isEmpty else { throw ImageHashError.emptyBits }
		let w = bits[0].count
		for row in bits where row.count != w {
			throw ImageHashError.nonRectangular
		}
		self.bits = bits
	}

	public var hex: String { Bitpack.pack(bits) }
	public var description: String { hex }

	public var bitCount: Int {
		bits.isEmpty ? 0 : bits.count * bits[0].count
	}

	public func subtract(_ other: Hash) throws -> Int {
		let h1 = bits.count, w1 = bits[0].count
		let h2 = other.bits.count, w2 = other.bits[0].count
		guard h1 == h2 && w1 == w2 else {
			throw ImageHashError.shapeMismatch(
				lhs: ImageHashError.ShapeKey(h1, w1),
				rhs: ImageHashError.ShapeKey(h2, w2)
			)
		}
		var diff = 0
		for y in 0..<h1 {
			for x in 0..<w1 where bits[y][x] != other.bits[y][x] {
				diff += 1
			}
		}
		return diff
	}
}

File: hash/js/rosetta-squint-hash/src/hash.ts
/**
 * Hash container, ImageHashError, and RgbImage interface.
 *
 * Public surface that consumers use directly. Algorithms create Hash
 * instances via `new Hash(bits)` after constructing the bit array.
 */

import { pack } from "./internal/bitpack.js";

export interface RgbImage {
  width: number;
  height: number;
  /** RGB or RGBA bytes, row-major. Length = width * height * channels. */
  data: Uint8Array;
  channels: 3 | 4;
}

export type ImageHashErrorKind =
  | "InvalidHashSize"
  | "NotPowerOfTwo"
  | "HashSizeTooLarge"
  | "InvalidBinbits"
  | "InvalidHex"
  | "ShapeMismatch"
  | "EmptyBits"
  | "NonRectangular";

export class ImageHashError extends Error {
  readonly kind: ImageHashErrorKind;
  constructor(kind: ImageHashErrorKind, message: string) {
    super(message);
    this.name = "ImageHashError";
    this.kind = kind;
  }
}

export class Hash {
  readonly #bits: boolean[][];

  constructor(bits: boolean[][]) {
    if (bits.length === 0 || bits[0].length === 0) {
      throw new ImageHashError("EmptyBits", "bits must be non-empty");
    }
    const w = bits[0].length;
    for (const row of bits) {
      if (row.length !== w) {
        throw new ImageHashError(
          "NonRectangular",
          `bits must be rectangular; got width ${row.length} vs ${w}`,
        );
      }
    }
    // Defensive copy so external mutation can't corrupt the hash
    this.#bits = bits.map(row => row.slice());
  }

  toHex(): string {
    return pack(this.#bits);
  }

  subtract(other: Hash): number {
    const h1 = this.#bits.length;
    const w1 = this.#bits[0].length;
    const o = other.#bits;
    const h2 = o.length;
    const w2 = o[0].length;
    if (h1 !== h2 || w1 !== w2) {
      throw new ImageHashError(
        "ShapeMismatch",
        `shapes don't match: this=(${h1},${w1}), other=(${h2},${w2})`,
      );
    }
    let diff = 0;
    for (let y = 0; y < h1; y++) {
      for (let x = 0; x < w1; x++) {
        if (this.#bits[y][x] !== o[y][x]) diff++;
      }
    }
    return diff;
  }

  bitCount(): number {
    if (this.#bits.length === 0) return 0;
    return this.#bits.length * this.#bits[0].length;
  }

  equals(other: Hash): boolean {
    if (this.#bits.length !== other.#bits.length) return false;
    for (let y = 0; y < this.#bits.length; y++) {
      const a = this.#bits[y];
      const b = other.#bits[y];
      if (a.length !== b.length) return false;
      for (let x = 0; x < a.length; x++) {
        if (a[x] !== b[x]) return false;
      }
    }
    return true;
  }

  toString(): string {
    return this.toHex();
  }
}
